ENG-11747: Add resource and datasource for new "log management" integration

gabfelp opened the pull request:
## Description of the change

There are new generic APIs that will be used for CRUDing all types of log integrations. The changes in this PR are meant to Support this new API in Terraform Provider.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklists

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing

Note: `enable_audit_logs` changed to `receive_audit_logs`, but the tests keep passing.

I created the respective resource:

```
resource "cyral_integration_logs" "int_logs" {
  name              = "myintegrationtest"
  enable_audit_logs = true
  config_scheme {
    elk {
      es_url     = "http://es.com"
      kibana_url = "http://kibana.com"
      es_credentials {
        username = "gabriel"
        password = "password"
      }
    }
  }
}
```

I ran `terraform init`and `terraform apply`

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # cyral_integration_logs.int_logs will be created
  + resource "cyral_integration_logs" "int_logs" {
      + enable_audit_logs = true
      + id                = (known after apply)
      + name              = "myintegrationtest"

      + config_scheme {
          + elk {
              + es_url     = "http://es.com"
              + kibana_url = "http://kibana.com"

              + es_credentials {
                  + password = "password"
                  + username = "gabriel"
                }
            }
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cyral_integration_logs.int_logs: Creating...
cyral_integration_logs.int_logs: Creation complete after 3s [id=XXXXXXXX]
```

After this I ran `terraform show`

```
# cyral_integration_logs.int_logs:
resource "cyral_integration_logs" "int_logs" {
    enable_audit_logs = true
    id                = "XXXXXXX"
    name              = "myintegrationtest"

    config_scheme {
        elk {
            es_url     = "http://es.com"
            kibana_url = "http://kibana.com"

            es_credentials {
                password = "password"
                username = "gabriel"
            }
        }
    }
}
```

By changing my tf file to:

```
resource "cyral_integration_logs" "int_logs" {
  name              = "gabriel-test"
  enable_audit_logs = true
  config_scheme {
    elk {
      es_url     = "http://es.com"
      kibana_url = "http://kibana.com"
      es_credentials {
        username = "another-user"
        password = "123"
      }
    }
  }
}
```

By running `terraform apply`

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # cyral_integration_logs.int_logs will be updated in-place
  ~ resource "cyral_integration_logs" "int_logs" {
        id                = "XXXXXXXX"
      ~ name              = "myintegrationtest" -> "gabriel-test"
        # (1 unchanged attribute hidden)

      ~ config_scheme {
          - elk {
              - es_url     = "http://es.com" -> null
              - kibana_url = "http://kibana.com" -> null

              - es_credentials {
                  - password = "password" -> null
                  - username = "gabriel" -> null
                }
            }
          + elk {
              + es_url     = "http://es.com"
              + kibana_url = "http://kibana.com"

              + es_credentials {
                  + password = "123"
                  + username = "another-user"
                }
            }
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cyral_integration_logs.int_logs: Modifying... [id=XXXXXXXXX]
cyral_integration_logs.int_logs: Modifications complete after 2s [id=XXXXXXXX]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

Result:

```
# cyral_integration_logs.int_logs:
resource "cyral_integration_logs" "int_logs" {
    enable_audit_logs = true
    id                = "XXXXXXXX"
    name              = "gabriel-test"

    config_scheme {
        elk {
            es_url     = "http://es.com"
            kibana_url = "http://kibana.com"

            es_credentials {
                password = "123"
                username = "another-user"
            }
        }
    }
}
```

This worked fine, so I tested the data source, by creating a file like:

```
data "cyral_integration_logs" "list" {
  type = "ELK"
}
```

After running `terraform apply` and `terraform show`:

```
data "cyral_integration_logs" "list" {
    id               = "6XXXX"
        {
            config_scheme     = [
                {
                    cloud_watch = []
                    datadog     = []
                    elk         = [
                        {
                            es_credentials = [
                                {
                                    password = "123"
                                    username = "another-user"
                                },
                            ]
                            es_url         = "http://es.com"
                            kibana_url     = "http://kibana.com"
                        },
                    ]
                    splunk      = []
                    sumo_logic  = []
                },
            ]
            enable_audit_logs = true
            id                = "XXXXXXX"
            name              = "gabriel-test"
        },
    ]
    type             = "ELK"
}
```

Then, for testing delete, I removed the `cyral_integration_logs` resource. 
`terraform apply`:

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # cyral_integration_logs.int_logs will be destroyed
  # (because cyral_integration_logs.int_logs is not in configuration)
  - resource "cyral_integration_logs" "int_logs" {
      - enable_audit_logs = true -> null
      - id                = "XXXXXXX" -> null
      - name              = "gabriel-test" -> null

      - config_scheme {
          - elk {
              - es_url     = "http://es.com" -> null
              - kibana_url = "http://kibana.com" -> null

              - es_credentials {
                  - password = "123" -> null
                  - username = "another-user" -> null
                }
            }
        }
    }

Plan: 0 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cyral_integration_logs.int_logs: Destroying... [id=XXXXXXXX]
cyral_integration_logs.int_logs: Destruction complete after 2s

Apply complete! Resources: 0 added, 0 changed, 1 destroyed.
```

After that, if we look for the data source list, the resource was gone.

VictorGFM commented on cyral/data_source_cyral_integration_logs_test.go:
```diff
@@ -0,0 +1,124 @@
+	testConfigCloudWatch, testFuncCloudWatch := setupLogsTest(initialLogsConfigCloudWatch)
+	testUpdateConfigCloudWatch, testUpdateFuncCloudWatch := setupLogsTest(updatedLogsConfigCloudWatch)
+
+	resource.ParallelTest(t, resource.TestCase{
```
Please make sure that these are not flaky tests and are independent of the existing resources in a CP. You can use the tests for the existing data sources as a reference.

gabfelp replied:
Thanks, I improved the datasource tests in here: [1c83973](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/1c839735700362291d95554a13a9e1995f341452)

VictorGFM reviewed (commented):
Thanks for working on this @gabfelp, I have some comments below.

wcmjunior commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+			return err
+		}
+
+		integrationList = append(integrationList, map[string]interface{}{
```
Should be `integrationsList`. Missed the plural. Maybe using just `integrations` makes more sense as this is not an actual list data structure.

gabfelp replied:
Addr here: [e7a72a9](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/e7a72a9c6616ff48fb3487ee132491a3900f7939)

wcmjunior commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+				Optional:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{
+					"ANY",
```
Why upper case? Do we store as upper case in the backend?

gabfelp replied:
Yes! As seen here: https://github.com/cyralinc/integrations/blob/main/shims/logging/logging.swagger.json#L45-L52

wcmjunior commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+									Type:        schema.TypeString,
+								},
+								"index": {
+									Description: "The Slunk index.",
```
```suggestion
									Description: "The Splunk index.",
```

gabfelp replied:
Addr here: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R239

wcmjunior reviewed (commented):
Partial review.

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+	"elk",
+	"splunk",
+	"sumo_logic",
+	"fluentbit",
```
Yes, this is a WIP by Chris https://github.com/cyralinc/integrations/pull/573

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+			Description: "Whether or not Cyral audit logs should be forwarded to this logging integration.",
+			Optional:    true,
+			Type:        schema.TypeBool,
+			Default:     true,
```
Confirmed with Chris that it is false. I'll be changing that, thanks!

gabfelp replied:
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R113

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+			MaxItems: 1,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"cloud_watch": {
```
Yes! I was working on this and here it is: [1848c1c](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/1848c1ca144abcf9ccd29d622057f3f1d856d583). Thanks!

gabfelp commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+					"ELK",
+					"SPLUNK",
+					"SUMOLOGIC",
+					"FLUENTBIT",
```
Yes it is, thanks for checking!

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+			Description: "Config option for your integration. List of supported types: " +
+				supportedTypesMarkdown(allLogIntegrationConfigs),
+			Required: true,
+			Type:     schema.TypeList,
```
Hmm, I tried changing this for `TypeSet` before and I got an error like this one:

```
1 error occurred:
│       * resource cyral_integration_logging: ExactlyOneOf: sumo_logic configuration block reference (config.0.cloud_watch) can only be used with TypeList and MaxItems: 1 configuration blocks
```

VictorGFM replied:
I see, these terraform schema constraints are odd and hard to memorize. Keeping this field as TypeList should be fine in this case then.

gabfelp commented on cyral/provider.go:
```diff
@@ -105,6 +106,7 @@ func Provider() *schema.Provider {
 			"cyral_integration_idp_saml":             resourceIntegrationIdPSAML(),
 			"cyral_integration_idp_saml_draft":       resourceIntegrationIdPSAMLDraft(),
 			"cyral_integration_sumo_logic":           resourceIntegrationSumoLogic(),
+			"cyral_integration_logs":                 resourceIntegrationLogs(),
```
Addr here: [provider](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-a86c00a3b0901ce990ddd38a99fa0644e7651c9fc44c091ae09ec1bf7a186147)

gabfelp commented on cyral/provider.go:
```diff
@@ -74,6 +74,7 @@ func Provider() *schema.Provider {
 			"cyral_datalabel":            dataSourceDatalabel(),
 			"cyral_integration_idp":      dataSourceIntegrationIdP(),
 			"cyral_integration_idp_saml": dataSourceIntegrationIdPSAML(),
+			"cyral_integration_logs":     dataSourceIntegrationLogs(),
```
Addr here: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-a86c00a3b0901ce990ddd38a99fa0644e7651c9fc44c091ae09ec1bf7a186147R77

gabfelp commented on cyral/resource_cyral_integration_logs.go:
```diff
@@ -0,0 +1,253 @@
+func resourceIntegrationLogs() *schema.Resource {
+
+	return &schema.Resource{
+		Description: "Manages Integrations with log services.",
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-d92f80b3b7bc610107cbe569d4ff2c14455201659f8451319cdc09b84b1e0607R236

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+			Type:        schema.TypeBool,
+			Default:     true,
+		},
+		"config_scheme": {
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R115 and changed all the places where this is used.

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+			Default:     true,
+		},
+		"config_scheme": {
+			Description: "Config option for your integration. List of supported types: " +
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R116

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+								},
+								"stream": {
+									Description: "The CloudWatch log stream.",
+									Required:    true,
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R143

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+									Type:        schema.TypeString,
+								},
+								"es_credentials": {
+									Description: "Object to comport Elastic Search credentials",
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R192

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+												Description: "The Elasticsearch password.",
+												Optional:    true,
+												Type:        schema.TypeString,
+											},
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R200 and https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R205

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+									Type:        schema.TypeString,
+								},
+								"index": {
+									Description: "The Slunk index.",
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R239

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+								},
+								"use_tls": {
+									Description: "Whether or not to use TLS.",
+									Required:    true,
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R245

gabfelp commented on cyral/resource_cyral_integration_logs.go:
```diff
@@ -0,0 +1,253 @@
+
+const loggingApiUrl = "https://%s/v1/integrations/logging/%s"
+
+func writeConfigScheme(resource *IntegrationLogConfig) ([]interface{}, error) {
```
Addr here https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-d92f80b3b7bc610107cbe569d4ff2c14455201659f8451319cdc09b84b1e0607R14 and in the other places where this is called

gabfelp commented on cyral/resource_cyral_integration_logs.go:
```diff
@@ -0,0 +1,253 @@
+				return fmt.Errorf(
+					"at most one elastic search credential is expected",
+				)
+			}
```
Yes, this makes sense, addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-d92f80b3b7bc610107cbe569d4ff2c14455201659f8451319cdc09b84b1e0607R155

gabfelp commented on cyral/utils.go:
```diff
@@ -113,3 +113,27 @@ func getStrList(m map[string]interface{}, attName string) []string {
+	} else {
+		return "false"
+	}
+}
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-a1bb606ca4f9df3c390d85f6eace0bf787c25de857f2856b545022eb0b223551R131 thanks for the tip! Removed from the other places

gabfelp commented on cyral/resource_cyral_integration_logs_test.go:
```diff
@@ -0,0 +1,232 @@
+	return fmt.Sprintf(`
+	resource "cyral_integration_logs" "logs_integration" {
+		name = "%s"
+		receive_audit_logs = "%s"
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+	Config string `json:"config"`
+}
+
+type IntegrationLogConfig struct {
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R47 and in the other places this is used

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+	IntegrationConfigScheme
+}
+
+type IntegrationConfigScheme struct {
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-3e52e669054f7a2a6a9b01f16bc7e5473a5c993332f5816f478f5990ad893b80R54 and in the other places

gabfelp commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+	computedSchema := schemaAllComputed(rawSchema)
+	log.Printf("[INFO] Computed schema: %v", computedSchema)
+	return &schema.Resource{
+		Description: "Retrieve and filter integrations.",
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-c7fc6a5d4a8c9648657cc0c25ac87653dc5e13cc550ad93bcb6c6b476425e027R64

gabfelp commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+		Description: "Retrieve and filter integrations.",
+		ReadContext: ReadResource(dataSourceIntegrationLogsRead()),
+		Schema: map[string]*schema.Schema{
+			"type": {
```
Addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-c7fc6a5d4a8c9648657cc0c25ac87653dc5e13cc550ad93bcb6c6b476425e027R80

gabfelp commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+
+func dataSourceIntegrationLogs() *schema.Resource {
+	rawSchema := getIntegrationLogsSchema()
+	// all fields in data_source are computed.
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-c7fc6a5d4a8c9648657cc0c25ac87653dc5e13cc550ad93bcb6c6b476425e027R59

gabfelp commented on cyral/data_source_cyral_integration_logs.go:
```diff
@@ -0,0 +1,95 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+type ListIntegrationLogsRequest struct {
```
Nowhere! this is not needed anymore! Thanks for catching this.

gabfelp commented on cyral/utils.go:
```diff
@@ -113,3 +113,27 @@ func getStrList(m map[string]interface{}, attName string) []string {
+		s[k].Computed = true
+		s[k].Default = nil
+		s[k].MaxItems = 0
+		s[k].ExactlyOneOf = nil
```
We thought about this and my concern at first was if these variables we're adding are the only ones that matter

In any case, if the list of parameters to update is longer than to add, it's better to go with your suggestion.

Thank you and addr: https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-a1bb606ca4f9df3c390d85f6eace0bf787c25de857f2856b545022eb0b223551R119-R124

gabfelp commented on cyral/utils.go:
```diff
@@ -113,3 +113,27 @@ func getStrList(m map[string]interface{}, attName string) []string {
 	return attStrs
 }
+
+func schemaAllComputed(s map[string]*schema.Schema) map[string]*schema.Schema {
```
Addr https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/5d305e608bafc5e59a8f1e846d68212cd4190d09#diff-a1bb606ca4f9df3c390d85f6eace0bf787c25de857f2856b545022eb0b223551R117 and in other places as well

gabfelp commented on cyral/model_integration_logs.go:
```diff
@@ -0,0 +1,264 @@
+									Type:        schema.TypeString,
+								},
+								"group": {
+									Description: "The CloudWatch log group.",
```
Addr here: [060d167](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/060d1679d87e4c8fc42d634b5aae98f356373a47)

gabfelp commented on cyral/resource_cyral_integration_logs_test.go:
```diff
@@ -0,0 +1,232 @@
+	},
+}
+
+func TestAccLogsIntegrationResourceElk(t *testing.T) {
```
Addr tests for all of them in here: [1c83973](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/1c839735700362291d95554a13a9e1995f341452)

gabfelp commented:
I tested after addressing the suggestions and it keeps working just fine:) 
![image](https://github.com/cyralinc/terraform-provider-cyral/assets/37452507/78b4d051-96d9-414d-8598-2590fae97be1)

VictorGFM commented on cyral/model_integration_logging.go:
```diff
@@ -108,20 +129,26 @@ func getIntegrationLogsSchema() map[string]*schema.Schema {
+									Description:  "Log retention days (optional). If specified, valid values are [1, 3, 5, 7, 14, 30, 60, 90, 120, 150, 180, 365, 400, 545, 731, 1827, 3653]. If unspecified, CloudWatch will retain the logs indefinitely.",
+									Optional:     true,
+									Type:         schema.TypeInt,
+									ValidateFunc: validateLogRetentionDays,
```
Would be better to delegate this validation to the API, since it already validates this field and if we ever need to change the validation in the future it would be easier to maintain it in a single place. I understand that adding the validation directly in the terraform scheme improves the resource UX a little bit, since the error will be caught earlier by the terraform instead of during the runtime of the API, so, in the end, is a tradeoff between a slight improvement in the UX vs maintenance, and I think in this case its better to opt for the latter. This means that the `validateLogRetentionDays` function can be removed.

gabfelp replied:
Addr [bb779d9](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/bb779d978dffb1a275a7d719ea74685ee6f3e497)

VictorGFM commented on cyral/data_source_cyral_integration_logging.go:
```diff
@@ -81,6 +77,7 @@ func dataSourceIntegrationLogs() *schema.Resource {
 				}, false),
+				Default: "ANY",
 			},
 			"integration_list": {
```
I know that some data sources use the list name in singular, but thinking better about that, I believe would be more correct to use plural here.
```suggestion
			"integrations_list": {
```

VictorGFM replied:
@wcmjunior any thoughts about that? Maybe just `integrations` would be better?

wcmjunior replied:
Absolutely.

VictorGFM reviewed (approved):
LGTM, please consider addressing my two comments below before merging

gabfelp commented:
I retested (manually and using the written tests) after addressing all suggestions and everything is working fine :) 

[
```
go clean -i github.com/cyralinc/terraform-provider-cyral/...
rm -f terraform-provider-cyral_v4.2.0
rm -rf ./out
rm -rf ~/.terraform.d/plugins/local/terraform/cyral/4.2.0
gofmt -w .
mkdir -p out/
# Build for both MacOS and Linux
GOOS=darwin GOARCH=amd64 go build -o out/darwin_amd64/terraform-provider-cyral_v4.2.0 .
GOOS=darwin GOARCH=arm64 go build -o out/darwin_arm64/terraform-provider-cyral_v4.2.0 .
GOOS=linux GOARCH=amd64 go build -o out/linux_amd64/terraform-provider-cyral_v4.2.0 .
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_amd64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_arm64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/linux_amd64
cp out/darwin_arm64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_arm64
cp out/linux_amd64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/linux_amd64
cp out/darwin_amd64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_amd64
go test github.com/cyralinc/terraform-provider-cyral/... -v -race -timeout 20m
?       github.com/cyralinc/terraform-provider-cyral    [no test files]
=== RUN   TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue
2023/05/18 14:36:07 [DEBUG] Init NewClient
2023/05/18 14:36:07 [DEBUG] TokenSource: &{0xc000012e88 {0 0} <nil>}
2023/05/18 14:36:07 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue (0.00s)
=== RUN   TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse
2023/05/18 14:36:07 [DEBUG] Init NewClient
2023/05/18 14:36:07 [DEBUG] TokenSource: &{0xc000012eb8 {0 0} <nil>}
2023/05/18 14:36:07 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse (0.00s)
=== RUN   TestNewClient_WhenClientIDIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientIDIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenClientSecretIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientSecretIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/client     (cached)

=== RUN   TestAccDatalabelDataSource
=== PAUSE TestAccDatalabelDataSource
=== RUN   TestAccIntegrationIdPSAMLDataSource
=== PAUSE TestAccIntegrationIdPSAMLDataSource
=== RUN   TestAccLoggingIntegrationDataSource
=== PAUSE TestAccLoggingIntegrationDataSource
=== RUN   TestAccRepositoryDataSource
=== PAUSE TestAccRepositoryDataSource
=== RUN   TestAccRoleDataSource
=== PAUSE TestAccRoleDataSource
=== RUN   TestAccSAMLCertificateDataSource
=== PAUSE TestAccSAMLCertificateDataSource
=== RUN   TestAccSAMLConfigurationDataSource
=== PAUSE TestAccSAMLConfigurationDataSource
=== RUN   TestAccSidecarBoundPortsDataSource
=== PAUSE TestAccSidecarBoundPortsDataSource
=== RUN   TestAccSidecarCftTemplateDataSource
=== PAUSE TestAccSidecarCftTemplateDataSource
=== RUN   TestAccSidecarIDDataSource
=== PAUSE TestAccSidecarIDDataSource
=== RUN   TestAccSidecarInstanceIDsDataSource
=== PAUSE TestAccSidecarInstanceIDsDataSource
=== RUN   TestIntegrationsData_GetValue_Default
--- PASS: TestIntegrationsData_GetValue_Default (0.00s)
=== RUN   TestIntegrationsData_GetValue_Splunk
--- PASS: TestIntegrationsData_GetValue_Splunk (0.00s)
=== RUN   TestAccProvider
--- PASS: TestAccProvider (0.00s)
=== RUN   TestAccDatalabelResource
=== PAUSE TestAccDatalabelResource
=== RUN   TestAccDatadogIntegrationResource
=== PAUSE TestAccDatadogIntegrationResource
=== RUN   TestAccELKIntegrationResource
=== PAUSE TestAccELKIntegrationResource
=== RUN   TestAccHCVaultIntegrationResource
=== PAUSE TestAccHCVaultIntegrationResource
=== RUN   TestAccIntegrationIdPSAMLDraftResource
=== PAUSE TestAccIntegrationIdPSAMLDraftResource
=== RUN   TestAccIntegrationIdPSAMLResource
=== PAUSE TestAccIntegrationIdPSAMLResource
=== RUN   TestAccIdPIntegrationResource
=== PAUSE TestAccIdPIntegrationResource
=== RUN   TestAccLogsIntegrationResourceCloudWatch
=== PAUSE TestAccLogsIntegrationResourceCloudWatch
=== RUN   TestAccLogsIntegrationResourceDataDog
=== PAUSE TestAccLogsIntegrationResourceDataDog
=== RUN   TestAccLogsIntegrationResourceElk
=== PAUSE TestAccLogsIntegrationResourceElk
=== RUN   TestAccLogsIntegrationResourceSplunk
=== PAUSE TestAccLogsIntegrationResourceSplunk
=== RUN   TestAccLogsIntegrationResourceSumologic
=== PAUSE TestAccLogsIntegrationResourceSumologic
=== RUN   TestAccLogsIntegrationResourceFluentbit
=== PAUSE TestAccLogsIntegrationResourceFluentbit
=== RUN   TestAccLogstashIntegrationResource
=== PAUSE TestAccLogstashIntegrationResource
=== RUN   TestAccLookerIntegrationResource
=== PAUSE TestAccLookerIntegrationResource
=== RUN   TestAccDuoMFAIntegrationResource
=== PAUSE TestAccDuoMFAIntegrationResource
=== RUN   TestAccPagerDutyIntegrationResource
=== PAUSE TestAccPagerDutyIntegrationResource
=== RUN   TestAccSlackAlertsIntegrationResource
=== PAUSE TestAccSlackAlertsIntegrationResource
=== RUN   TestAccSplunkIntegrationResource
=== PAUSE TestAccSplunkIntegrationResource
=== RUN   TestAccSumoLogicIntegrationResource
=== PAUSE TestAccSumoLogicIntegrationResource
=== RUN   TestAccMsTeamsIntegrationResource
=== PAUSE TestAccMsTeamsIntegrationResource
=== RUN   TestAccPolicyRuleResource
=== PAUSE TestAccPolicyRuleResource
=== RUN   TestPolicyRuleResourceUpgradeV0
--- PASS: TestPolicyRuleResourceUpgradeV0 (0.00s)
=== RUN   TestAccPolicyResource
=== PAUSE TestAccPolicyResource
=== RUN   TestAccRepositoryAccessGatewayResource
=== PAUSE TestAccRepositoryAccessGatewayResource
=== RUN   TestAccRepositoryAccessRulesResource
=== PAUSE TestAccRepositoryAccessRulesResource
=== RUN   TestAccRepositoryBindingResource
=== PAUSE TestAccRepositoryBindingResource
=== RUN   TestAccRepositoryConfAnalysisResource
=== PAUSE TestAccRepositoryConfAnalysisResource
=== RUN   TestRepositoryConfAnalysisResourceUpgradeV0
--- PASS: TestRepositoryConfAnalysisResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryConfAuthResource
=== PAUSE TestAccRepositoryConfAuthResource
=== RUN   TestRepositoryConfAuthResourceUpgradeV0
--- PASS: TestRepositoryConfAuthResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryDatamapResource
=== PAUSE TestAccRepositoryDatamapResource
=== RUN   TestAccRepositoryNetworkAccessPolicyResource
=== PAUSE TestAccRepositoryNetworkAccessPolicyResource
=== RUN   TestAccRepositoryResource
=== PAUSE TestAccRepositoryResource
=== RUN   TestAccRepositoryUserAccountResource
=== PAUSE TestAccRepositoryUserAccountResource
=== RUN   TestAccRoleSSOGroupsResource
=== PAUSE TestAccRoleSSOGroupsResource
=== RUN   TestRoleSSOGroupsResourceUpgradeV0
--- PASS: TestRoleSSOGroupsResourceUpgradeV0 (0.00s)
=== RUN   TestAccRoleResource
=== PAUSE TestAccRoleResource
=== RUN   TestAccSidecarCredentialsResource
=== PAUSE TestAccSidecarCredentialsResource
=== RUN   TestSidecarListenerResource
=== PAUSE TestSidecarListenerResource
=== RUN   TestAccSidecarResource
=== PAUSE TestAccSidecarResource
=== RUN   TestElementsMatch
--- PASS: TestElementsMatch (0.00s)
=== CONT  TestAccSAMLCertificateDataSource
=== CONT  TestAccLogsIntegrationResourceFluentbit
=== CONT  TestAccDatadogIntegrationResource
=== CONT  TestAccIdPIntegrationResource
=== CONT  TestAccLogsIntegrationResourceSplunk
=== CONT  TestAccLogsIntegrationResourceElk
=== CONT  TestAccLogsIntegrationResourceSumologic
=== CONT  TestAccSidecarCftTemplateDataSource
--- PASS: TestAccSAMLCertificateDataSource (6.04s)
=== CONT  TestAccLogsIntegrationResourceDataDog
=== CONT  TestAccLookerIntegrationResource
--- PASS: TestAccLogsIntegrationResourceFluentbit (7.65s)
--- PASS: TestAccDatadogIntegrationResource (8.25s)
=== CONT  TestAccRoleDataSource
--- PASS: TestAccLogsIntegrationResourceSumologic (8.72s)
=== CONT  TestAccDatalabelResource
--- PASS: TestAccLogsIntegrationResourceElk (8.74s)
=== CONT  TestAccLoggingIntegrationDataSource
--- PASS: TestAccLogsIntegrationResourceSplunk (8.76s)
=== CONT  TestAccRepositoryDataSource
=== CONT  TestAccSidecarBoundPortsDataSource
--- PASS: TestAccLogsIntegrationResourceDataDog (6.43s)
--- PASS: TestAccRoleDataSource (5.84s)
=== CONT  TestAccIntegrationIdPSAMLResource
=== CONT  TestAccHCVaultIntegrationResource
--- PASS: TestAccLookerIntegrationResource (7.33s)
=== CONT  TestAccSAMLConfigurationDataSource
--- PASS: TestAccSidecarCftTemplateDataSource (14.98s)
=== CONT  TestAccIntegrationIdPSAMLDraftResource
--- PASS: TestAccDatalabelResource (7.97s)
--- PASS: TestAccLoggingIntegrationDataSource (10.81s)
=== CONT  TestAccLogsIntegrationResourceCloudWatch
--- PASS: TestAccHCVaultIntegrationResource (7.57s)
=== CONT  TestAccSidecarIDDataSource
--- PASS: TestAccRepositoryDataSource (15.28s)
=== CONT  TestAccRepositoryDatamapResource
=== CONT  TestAccSidecarResource
--- PASS: TestAccSAMLConfigurationDataSource (9.59s)
--- PASS: TestAccSidecarBoundPortsDataSource (13.64s)
=== CONT  TestAccIntegrationIdPSAMLDataSource
=== CONT  TestAccRoleResource
--- PASS: TestAccLogsIntegrationResourceCloudWatch (7.50s)
--- PASS: TestAccIntegrationIdPSAMLDraftResource (10.39s)
=== CONT  TestSidecarListenerResource
--- PASS: TestAccSidecarIDDataSource (6.91s)
=== CONT  TestAccRepositoryUserAccountResource
--- PASS: TestAccIntegrationIdPSAMLResource (20.56s)
=== CONT  TestAccRepositoryConfAnalysisResource
=== CONT  TestAccRepositoryBindingResource
--- PASS: TestAccIdPIntegrationResource (42.89s)
=== CONT  TestAccRoleSSOGroupsResource
--- PASS: TestAccRepositoryConfAnalysisResource (9.68s)
--- PASS: TestAccIntegrationIdPSAMLDataSource (18.78s)
=== CONT  TestAccRepositoryAccessRulesResource
=== CONT  TestAccSplunkIntegrationResource
--- PASS: TestAccRepositoryDatamapResource (24.76s)
--- PASS: TestAccRoleResource (23.59s)
=== CONT  TestAccRepositoryAccessGatewayResource
--- PASS: TestAccSidecarResource (29.84s)
=== CONT  TestAccELKIntegrationResource
=== CONT  TestAccSlackAlertsIntegrationResource
--- PASS: TestAccRepositoryBindingResource (11.58s)
--- PASS: TestAccSplunkIntegrationResource (7.27s)
=== CONT  TestAccPagerDutyIntegrationResource
--- PASS: TestAccRoleSSOGroupsResource (12.30s)
=== CONT  TestAccLogstashIntegrationResource
--- PASS: TestAccRepositoryAccessRulesResource (14.99s)
=== CONT  TestAccDatalabelDataSource
=== CONT  TestAccRepositoryResource
--- PASS: TestAccSlackAlertsIntegrationResource (6.04s)
=== CONT  TestAccSidecarCredentialsResource
--- PASS: TestAccELKIntegrationResource (6.89s)
=== CONT  TestAccRepositoryNetworkAccessPolicyResource
--- PASS: TestSidecarListenerResource (35.33s)
=== CONT  TestAccRepositoryConfAuthResource
--- PASS: TestAccPagerDutyIntegrationResource (6.85s)
--- PASS: TestAccSidecarCredentialsResource (5.34s)
=== CONT  TestAccSumoLogicIntegrationResource
--- PASS: TestAccLogstashIntegrationResource (12.51s)
=== CONT  TestAccPolicyRuleResource
--- PASS: TestAccRepositoryAccessGatewayResource (19.77s)
=== CONT  TestAccDuoMFAIntegrationResource
=== CONT  TestAccMsTeamsIntegrationResource
--- PASS: TestAccSumoLogicIntegrationResource (6.72s)
=== CONT  TestAccPolicyResource
--- PASS: TestAccDuoMFAIntegrationResource (6.89s)
--- PASS: TestAccRepositoryUserAccountResource (47.93s)
=== CONT  TestAccSidecarInstanceIDsDataSource
--- PASS: TestAccRepositoryConfAuthResource (15.42s)
--- PASS: TestAccDatalabelDataSource (19.08s)
--- PASS: TestAccMsTeamsIntegrationResource (5.88s)
--- PASS: TestAccRepositoryResource (19.98s)
--- PASS: TestAccPolicyResource (6.84s)
--- PASS: TestAccRepositoryNetworkAccessPolicyResource (22.02s)
--- PASS: TestAccSidecarInstanceIDsDataSource (7.06s)
--- PASS: TestAccPolicyRuleResource (18.12s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral      87.326s
```
]

@wcmjunior if we are satisfied with the PR, could you merge it? I don't have permission for that 😓

wcmjunior commented on cyral/model_integration_logging.go:
```diff
@@ -0,0 +1,270 @@
+								},
+								"access_token": {
+									Description: "Splunk access token.",
+									Required:    true,
```
```suggestion
									Required:    true,
									Sensitive:    true,
```

wcmjunior commented on cyral/model_integration_logging.go:
```diff
@@ -0,0 +1,270 @@
+											},
+											"password": {
+												Description: "Elasticsearch password.",
+												Required:    true,
```
```suggestion
												Required:    true,
												Sensitive:    true,
```

wcmjunior commented on cyral/model_integration_logging.go:
```diff
@@ -0,0 +1,270 @@
+							Schema: map[string]*schema.Schema{
+								"api_key": {
+									Description: "DataDog API key.",
+									Required:    true,
```
```suggestion
									Required:    true,
									Sensitive:    true,
```

gabfelp commented:
I address the changes for removing "config" from resource. Instead, we will be using the ConflictsWith for adding mutual exclusion logic.

I made a new test using:

```
resource "cyral_integration_logging" "int_logs_elk" {
  name               = "gabriel-test_elk"
  receive_audit_logs = true
  elk {
    es_url     = "http://es.com"
    kibana_url = "http://kibana.com"
    es_credentials {
      username = "another-user"
      password = "123"
    }
  }
}

data "cyral_integration_logging" "list" {
  type = "CLOUDWATCH"
}

resource "cyral_integration_logging" "int_logs_cloudd" {
  name               = "gabriel-test_cloud"
  receive_audit_logs = true
  cloud_watch {
    region = "us-east-1"
    group  = "everyone"
    stream = "abc"
  }
}

resource "cyral_integration_logging" "int_logs_datadog_2" {
  name               = "gabriel-test_datadog"
  receive_audit_logs = true
  datadog {
    api_key = "ANYTHING_IS_API_KEY"
  }
}

resource "cyral_integration_logging" "int_logs_fluentbit" {
  name               = "gabriel-fluent_bit"
  receive_audit_logs = true
  fluentbit {
    config = <<-EOF
      [OUTPUT]
      Name         stdout
      Match        *
    EOF
  }
}
```

After `terraform init` and `terraform apply`:

```
data.cyral_integration_logging.list: Reading...
data.cyral_integration_logging.list: Read complete after 2s [id=XXXXXXXXX]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # cyral_integration_logging.int_logs_cloudd will be created
  + resource "cyral_integration_logging" "int_logs_cloudd" {
      + id                 = (known after apply)
      + name               = "gabriel-test_cloud"
      + receive_audit_logs = true

      + cloud_watch {
          + group  = "everyone"
          + region = "us-east-1"
          + stream = "abc"
        }
    }

  # cyral_integration_logging.int_logs_datadog_2 will be created
  + resource "cyral_integration_logging" "int_logs_datadog_2" {
      + id                 = (known after apply)
      + name               = "gabriel-test_datadog"
      + receive_audit_logs = true

      + datadog {
          # At least one attribute in this block is (or was) sensitive,
          # so its contents will not be displayed.
        }
    }

  # cyral_integration_logging.int_logs_elk will be created
  + resource "cyral_integration_logging" "int_logs_elk" {
      + id                 = (known after apply)
      + name               = "gabriel-test_elk"
      + receive_audit_logs = true

      + elk {
          # At least one attribute in this block is (or was) sensitive,
          # so its contents will not be displayed.
        }
    }

  # cyral_integration_logging.int_logs_fluentbit will be created
  + resource "cyral_integration_logging" "int_logs_fluentbit" {
      + id                 = (known after apply)
      + name               = "gabriel-fluent_bit"
      + receive_audit_logs = true

      + fluentbit {
          + config = <<-EOT
                [OUTPUT]
                Name         stdout
                Match        *
            EOT
        }
    }

Plan: 4 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cyral_integration_logging.int_logs_cloudd: Creating...
cyral_integration_logging.int_logs_datadog_2: Creating...
cyral_integration_logging.int_logs_fluentbit: Creating...
cyral_integration_logging.int_logs_elk: Creating...
cyral_integration_logging.int_logs_fluentbit: Creation complete after 2s [id=XXXX]
cyral_integration_logging.int_logs_datadog_2: Creation complete after 2s [id=XXX]
cyral_integration_logging.int_logs_elk: Creation complete after 2s [id=XXXX]
cyral_integration_logging.int_logs_cloudd: Creation complete after 2s [id=XXX]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
```

By using `terraform show`:

```
data "cyral_integration_logging" "list" {
    id           = "XXXXX"
    integrations = [
        {
            cloud_watch        = [
                {
                    group              = "everyone"
                    log_retention_days = 0
                    region             = "us-east-1"
                    stream             = "abc"
                },
            ]
            datadog            = []
            elk                = []
            fluentbit          = []
            id                 = "XXXXXXXXX"
            name               = "gabriel-test"
            receive_audit_logs = true
            splunk             = []
            sumo_logic         = []
        },
        {
            cloud_watch        = [
                {
                    group              = "everyone"
                    log_retention_days = 0
                    region             = "us-east-1"
                    stream             = "abc"
                },
            ]
            datadog            = []
            elk                = []
            fluentbit          = []
            id                 = "XXXXXXX"
            name               = "gabriel-test22"
            receive_audit_logs = true
            splunk             = []
            sumo_logic         = []
        },
        {
            cloud_watch        = [
                {
                    group              = "everyone"
                    log_retention_days = 0
                    region             = "us-east-1"
                    stream             = "abc"
                },
            ]
            datadog            = []
            elk                = []
            fluentbit          = []
            id                 = "XXXXXX"
            name               = "gabriel-test23"
            receive_audit_logs = true
            splunk             = []
            sumo_logic         = []
        },
        {
            cloud_watch        = [
                {
                    group              = "group2"
                    log_retention_days = 1
                    region             = "us-east-2"
                    stream             = "abcd"
                },
            ]
            datadog            = []
            elk                = []
            fluentbit          = []
            id                 = "XXXXX"
            name               = "tfprov-acc-integration-log-LogsCloudWatchTest"
            receive_audit_logs = true
            splunk             = []
            sumo_logic         = []
        },
        {
            cloud_watch        = [
                {
                    group              = ""
                    log_retention_days = 0
                    region             = ""
                    stream             = "cyral-sidecar"
                },
            ]
            datadog            = []
            elk                = []
            fluentbit          = []
            id                 = "cloudwatch"
            name               = "AWS CloudWatch (default)"
            receive_audit_logs = false
            splunk             = []
            sumo_logic         = []
        },
    ]
    type         = "CLOUDWATCH"
}

# cyral_integration_logging.int_logs_cloudd:
resource "cyral_integration_logging" "int_logs_cloudd" {
    id                 = "XXXXXXX"
    name               = "gabriel-test_cloud"
    receive_audit_logs = true

    cloud_watch {
        group  = "everyone"
        region = "us-east-1"
        stream = "abc"
    }
}

# cyral_integration_logging.int_logs_datadog_2:
resource "cyral_integration_logging" "int_logs_datadog_2" {
    id                 = "XXXXXXX"
    name               = "gabriel-test_datadog"
    receive_audit_logs = true

    datadog {
      # At least one attribute in this block is (or was) sensitive,
      # so its contents will not be displayed.
    }
}

# cyral_integration_logging.int_logs_elk:
resource "cyral_integration_logging" "int_logs_elk" {
    id                 = "XXXXXXXX"
    name               = "gabriel-test_elk"
    receive_audit_logs = true

    elk {
      # At least one attribute in this block is (or was) sensitive,
      # so its contents will not be displayed.
    }
}

# cyral_integration_logging.int_logs_fluentbit:
resource "cyral_integration_logging" "int_logs_fluentbit" {
    id                 = "XXXXXXX"
    name               = "gabriel-fluent_bit"
    receive_audit_logs = true

    fluentbit {
        config = <<-EOT
            [OUTPUT]
            Name         stdout
            Match        *
        EOT
    }
}
```

(I had old CloudWatch integrations and that's why the Data Source returned a lot of integrations).

After removing things from the `.tf` file and running `terraform apply`, it removed the resource just fine

```
cyral_integration_logging.int_logs_fluentbit: Refreshing state... [id=XXXXXXX]
data.cyral_integration_logging.list: Reading...
cyral_integration_logging.int_logs_datadog_2: Refreshing state... [id=XXXXXXXX]
cyral_integration_logging.int_logs_elk: Refreshing state... [id=XXXXXX]
cyral_integration_logging.int_logs_cloudd: Refreshing state... [id=XXXXXXX]
data.cyral_integration_logging.list: Read complete after 2s [id=XXXXXXx]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # cyral_integration_logging.int_logs_fluentbit will be destroyed
  # (because cyral_integration_logging.int_logs_fluentbit is not in configuration)
  - resource "cyral_integration_logging" "int_logs_fluentbit" {
      - id                 = "XXXX" -> null
      - name               = "gabriel-fluent_bit" -> null
      - receive_audit_logs = true -> null

      - fluentbit {
          - config = <<-EOT
                [OUTPUT]
                Name         stdout
                Match        *
            EOT -> null
        }
    }

Plan: 0 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

cyral_integration_logging.int_logs_fluentbit: Destroying... [id=XXXXX]
cyral_integration_logging.int_logs_fluentbit: Destruction complete after 2s
`´`


Here's the automated test suite:

```
go clean -i github.com/cyralinc/terraform-provider-cyral/...
rm -f terraform-provider-cyral_v4.2.0
rm -rf ./out
rm -rf ~/.terraform.d/plugins/local/terraform/cyral/4.2.0
gofmt -w .
mkdir -p out/
# Build for both MacOS and Linux
GOOS=darwin GOARCH=amd64 go build -o out/darwin_amd64/terraform-provider-cyral_v4.2.0 .
GOOS=darwin GOARCH=arm64 go build -o out/darwin_arm64/terraform-provider-cyral_v4.2.0 .
GOOS=linux GOARCH=amd64 go build -o out/linux_amd64/terraform-provider-cyral_v4.2.0 .
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_amd64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_arm64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/linux_amd64
cp out/darwin_arm64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_arm64
cp out/linux_amd64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/linux_amd64
cp out/darwin_amd64/terraform-provider-cyral_v4.2.0 ~/.terraform.d/plugins/local/terraform/cyral/4.2.0/darwin_amd64
go test github.com/cyralinc/terraform-provider-cyral/... -v -race -timeout 20m
?       github.com/cyralinc/terraform-provider-cyral    [no test files]
=== RUN   TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue
2023/05/18 14:36:07 [DEBUG] Init NewClient
2023/05/18 14:36:07 [DEBUG] TokenSource: &{0xc000012e88 {0 0} <nil>}
2023/05/18 14:36:07 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue (0.00s)
=== RUN   TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse
2023/05/18 14:36:07 [DEBUG] Init NewClient
2023/05/18 14:36:07 [DEBUG] TokenSource: &{0xc000012eb8 {0 0} <nil>}
2023/05/18 14:36:07 [DEBUG] End NewClient
--- PASS: TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse (0.00s)
=== RUN   TestNewClient_WhenClientIDIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientIDIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenClientSecretIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenClientSecretIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError
2023/05/18 14:36:07 [DEBUG] Init NewClient
--- PASS: TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/client     (cached)
=== RUN   TestAccDatalabelDataSource
=== PAUSE TestAccDatalabelDataSource
=== RUN   TestAccIntegrationIdPSAMLDataSource
=== PAUSE TestAccIntegrationIdPSAMLDataSource
=== RUN   TestAccLoggingIntegrationDataSource
=== PAUSE TestAccLoggingIntegrationDataSource
=== RUN   TestAccRepositoryDataSource
=== PAUSE TestAccRepositoryDataSource
=== RUN   TestAccRoleDataSource
=== PAUSE TestAccRoleDataSource
=== RUN   TestAccSAMLCertificateDataSource
=== PAUSE TestAccSAMLCertificateDataSource
=== RUN   TestAccSAMLConfigurationDataSource
=== PAUSE TestAccSAMLConfigurationDataSource
=== RUN   TestAccSidecarBoundPortsDataSource
=== PAUSE TestAccSidecarBoundPortsDataSource
=== RUN   TestAccSidecarCftTemplateDataSource
=== PAUSE TestAccSidecarCftTemplateDataSource
=== RUN   TestAccSidecarIDDataSource
=== PAUSE TestAccSidecarIDDataSource
=== RUN   TestAccSidecarInstanceIDsDataSource
=== PAUSE TestAccSidecarInstanceIDsDataSource
=== RUN   TestIntegrationsData_GetValue_Default
--- PASS: TestIntegrationsData_GetValue_Default (0.00s)
=== RUN   TestIntegrationsData_GetValue_Splunk
--- PASS: TestIntegrationsData_GetValue_Splunk (0.00s)
=== RUN   TestAccProvider
--- PASS: TestAccProvider (0.00s)
=== RUN   TestAccDatalabelResource
=== PAUSE TestAccDatalabelResource
=== RUN   TestAccDatadogIntegrationResource
=== PAUSE TestAccDatadogIntegrationResource
=== RUN   TestAccELKIntegrationResource
=== PAUSE TestAccELKIntegrationResource
=== RUN   TestAccHCVaultIntegrationResource
=== PAUSE TestAccHCVaultIntegrationResource
=== RUN   TestAccIntegrationIdPSAMLDraftResource
=== PAUSE TestAccIntegrationIdPSAMLDraftResource
=== RUN   TestAccIntegrationIdPSAMLResource
=== PAUSE TestAccIntegrationIdPSAMLResource
=== RUN   TestAccIdPIntegrationResource
=== PAUSE TestAccIdPIntegrationResource
=== RUN   TestAccLogsIntegrationResourceCloudWatch
=== PAUSE TestAccLogsIntegrationResourceCloudWatch
=== RUN   TestAccLogsIntegrationResourceDataDog
=== PAUSE TestAccLogsIntegrationResourceDataDog
=== RUN   TestAccLogsIntegrationResourceElk
=== PAUSE TestAccLogsIntegrationResourceElk
=== RUN   TestAccLogsIntegrationResourceSplunk
=== PAUSE TestAccLogsIntegrationResourceSplunk
=== RUN   TestAccLogsIntegrationResourceSumologic
=== PAUSE TestAccLogsIntegrationResourceSumologic
=== RUN   TestAccLogsIntegrationResourceFluentbit
=== PAUSE TestAccLogsIntegrationResourceFluentbit
=== RUN   TestAccLogstashIntegrationResource
=== PAUSE TestAccLogstashIntegrationResource
=== RUN   TestAccLookerIntegrationResource
=== PAUSE TestAccLookerIntegrationResource
=== RUN   TestAccDuoMFAIntegrationResource
=== PAUSE TestAccDuoMFAIntegrationResource
=== RUN   TestAccPagerDutyIntegrationResource
=== PAUSE TestAccPagerDutyIntegrationResource
=== RUN   TestAccSlackAlertsIntegrationResource
=== PAUSE TestAccSlackAlertsIntegrationResource
=== RUN   TestAccSplunkIntegrationResource
=== PAUSE TestAccSplunkIntegrationResource
=== RUN   TestAccSumoLogicIntegrationResource
=== PAUSE TestAccSumoLogicIntegrationResource
=== RUN   TestAccMsTeamsIntegrationResource
=== PAUSE TestAccMsTeamsIntegrationResource
=== RUN   TestAccPolicyRuleResource
=== PAUSE TestAccPolicyRuleResource
=== RUN   TestPolicyRuleResourceUpgradeV0
--- PASS: TestPolicyRuleResourceUpgradeV0 (0.00s)
=== RUN   TestAccPolicyResource
=== PAUSE TestAccPolicyResource
=== RUN   TestAccRepositoryAccessGatewayResource
=== PAUSE TestAccRepositoryAccessGatewayResource
=== RUN   TestAccRepositoryAccessRulesResource
=== PAUSE TestAccRepositoryAccessRulesResource
=== RUN   TestAccRepositoryBindingResource
=== PAUSE TestAccRepositoryBindingResource
=== RUN   TestAccRepositoryConfAnalysisResource
=== PAUSE TestAccRepositoryConfAnalysisResource
=== RUN   TestRepositoryConfAnalysisResourceUpgradeV0
--- PASS: TestRepositoryConfAnalysisResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryConfAuthResource
=== PAUSE TestAccRepositoryConfAuthResource
=== RUN   TestRepositoryConfAuthResourceUpgradeV0
--- PASS: TestRepositoryConfAuthResourceUpgradeV0 (0.00s)
=== RUN   TestAccRepositoryDatamapResource
=== PAUSE TestAccRepositoryDatamapResource
=== RUN   TestAccRepositoryNetworkAccessPolicyResource
=== PAUSE TestAccRepositoryNetworkAccessPolicyResource
=== RUN   TestAccRepositoryResource
=== PAUSE TestAccRepositoryResource
=== RUN   TestAccRepositoryUserAccountResource
=== PAUSE TestAccRepositoryUserAccountResource
=== RUN   TestAccRoleSSOGroupsResource
=== PAUSE TestAccRoleSSOGroupsResource
=== RUN   TestRoleSSOGroupsResourceUpgradeV0
--- PASS: TestRoleSSOGroupsResourceUpgradeV0 (0.00s)
=== RUN   TestAccRoleResource
=== PAUSE TestAccRoleResource
=== RUN   TestAccSidecarCredentialsResource
=== PAUSE TestAccSidecarCredentialsResource
=== RUN   TestSidecarListenerResource
=== PAUSE TestSidecarListenerResource
=== RUN   TestAccSidecarResource
=== PAUSE TestAccSidecarResource
=== RUN   TestElementsMatch
--- PASS: TestElementsMatch (0.00s)
=== CONT  TestAccLoggingIntegrationDataSource
=== CONT  TestAccLogsIntegrationResourceFluentbit
=== CONT  TestAccRepositoryBindingResource
=== CONT  TestAccDatalabelResource
=== CONT  TestAccSumoLogicIntegrationResource
=== CONT  TestAccRepositoryAccessRulesResource
=== CONT  TestAccSAMLConfigurationDataSource
=== CONT  TestAccRepositoryAccessGatewayResource
=== CONT  TestAccDatadogIntegrationResource
--- PASS: TestAccSumoLogicIntegrationResource (8.65s)
=== CONT  TestAccIdPIntegrationResource
--- PASS: TestAccLogsIntegrationResourceFluentbit (8.79s)
--- PASS: TestAccDatalabelResource (9.13s)
=== CONT  TestAccSidecarInstanceIDsDataSource
=== CONT  TestAccMsTeamsIntegrationResource
--- PASS: TestAccSAMLConfigurationDataSource (11.12s)
=== CONT  TestAccLookerIntegrationResource
--- PASS: TestAccLoggingIntegrationDataSource (12.37s)
=== CONT  TestAccSAMLCertificateDataSource
--- PASS: TestAccRepositoryBindingResource (12.38s)
--- PASS: TestAccRepositoryAccessRulesResource (16.60s)
=== CONT  TestAccSidecarBoundPortsDataSource
=== CONT  TestAccRoleDataSource
--- PASS: TestAccDatadogIntegrationResource (8.17s)
--- PASS: TestAccSidecarInstanceIDsDataSource (7.73s)
=== CONT  TestAccIntegrationIdPSAMLResource
--- PASS: TestAccMsTeamsIntegrationResource (6.24s)
=== CONT  TestAccRoleResource
=== CONT  TestAccHCVaultIntegrationResource
--- PASS: TestAccSAMLCertificateDataSource (5.05s)
--- PASS: TestAccLookerIntegrationResource (7.68s)
=== CONT  TestAccRepositoryUserAccountResource
=== CONT  TestAccRepositoryConfAuthResource
--- PASS: TestAccRepositoryAccessGatewayResource (23.02s)
--- PASS: TestAccRoleDataSource (7.01s)
=== CONT  TestAccIntegrationIdPSAMLDraftResource
--- PASS: TestAccHCVaultIntegrationResource (7.94s)
=== CONT  TestAccRepositoryNetworkAccessPolicyResource
--- PASS: TestAccSidecarBoundPortsDataSource (13.71s)
=== CONT  TestAccRepositoryDatamapResource
=== CONT  TestAccLogsIntegrationResourceCloudWatch
--- PASS: TestAccIntegrationIdPSAMLDraftResource (11.46s)
--- PASS: TestAccRoleResource (22.02s)
=== CONT  TestAccRepositoryConfAnalysisResource
=== CONT  TestAccRoleSSOGroupsResource
--- PASS: TestAccIntegrationIdPSAMLResource (23.10s)
=== CONT  TestAccSlackAlertsIntegrationResource
--- PASS: TestAccRepositoryConfAuthResource (18.40s)
--- PASS: TestAccLogsIntegrationResourceCloudWatch (8.05s)
=== CONT  TestAccSidecarIDDataSource
--- PASS: TestAccSlackAlertsIntegrationResource (6.11s)
=== CONT  TestAccPolicyResource
=== CONT  TestAccPagerDutyIntegrationResource
--- PASS: TestAccRepositoryConfAnalysisResource (9.15s)
--- PASS: TestAccRepositoryNetworkAccessPolicyResource (23.74s)
=== CONT  TestAccSplunkIntegrationResource
--- PASS: TestAccSidecarIDDataSource (6.63s)
=== CONT  TestAccDuoMFAIntegrationResource
--- PASS: TestAccIdPIntegrationResource (43.85s)
=== CONT  TestAccLogstashIntegrationResource
=== CONT  TestAccSidecarCftTemplateDataSource
--- PASS: TestAccRoleSSOGroupsResource (12.97s)
=== CONT  TestAccRepositoryDataSource
--- PASS: TestAccRepositoryDatamapResource (23.72s)
--- PASS: TestAccPolicyResource (7.04s)
=== CONT  TestAccLogsIntegrationResourceSumologic
--- PASS: TestAccPagerDutyIntegrationResource (7.04s)
=== CONT  TestAccLogsIntegrationResourceSplunk
--- PASS: TestAccSplunkIntegrationResource (6.97s)
=== CONT  TestAccIntegrationIdPSAMLDataSource
--- PASS: TestAccDuoMFAIntegrationResource (7.13s)
=== CONT  TestAccPolicyRuleResource
--- PASS: TestAccLogsIntegrationResourceSumologic (7.76s)
=== CONT  TestSidecarListenerResource
=== CONT  TestAccELKIntegrationResource
--- PASS: TestAccLogsIntegrationResourceSplunk (7.80s)
--- PASS: TestAccSidecarCftTemplateDataSource (11.33s)
=== CONT  TestAccLogsIntegrationResourceDataDog
=== CONT  TestAccDatalabelDataSource
--- PASS: TestAccLogstashIntegrationResource (13.65s)
--- PASS: TestAccRepositoryDataSource (14.89s)
=== CONT  TestAccSidecarCredentialsResource
--- PASS: TestAccLogsIntegrationResourceDataDog (6.28s)
=== CONT  TestAccSidecarResource
=== CONT  TestAccLogsIntegrationResourceElk
--- PASS: TestAccRepositoryUserAccountResource (51.06s)
--- PASS: TestAccELKIntegrationResource (7.84s)
=== CONT  TestAccRepositoryResource
--- PASS: TestAccIntegrationIdPSAMLDataSource (16.08s)
--- PASS: TestAccSidecarCredentialsResource (5.46s)
--- PASS: TestAccPolicyRuleResource (19.37s)
--- PASS: TestAccLogsIntegrationResourceElk (7.14s)
--- PASS: TestAccDatalabelDataSource (18.88s)
--- PASS: TestAccRepositoryResource (21.07s)
--- PASS: TestSidecarListenerResource (32.76s)
--- PASS: TestAccSidecarResource (26.63s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral      97.245s
```

Thanks for the suggestion, @wcmjunior. Please let me know if you have any additional comments!

ccampo133 reviewed (changes requested):
@gabfelp we ended up removing log retention days for CloudWatch. Can you please remove that from the resource in this PR? Sorry for the churn.

gabfelp commented:
@ccampo133 removed this here: [56e769e](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/56e769e403ceb6d41fec1b1d7a727894dec37955)

wcmjunior commented on docs/resources/integration_logging.md:
```diff
@@ -0,0 +1,107 @@
+
+Optional:
+
+- `stream` (String) CloudWatch log stream.
```
I think it defaults to `cyral-sidecar` if not provided. @ccampo133 , could you please confirm? @gabfelp , please add this info to the docs once we get the confirmation.

gabfelp replied:
Sure, updated the description accordingly: [35daafa](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/35daafa7a278976ed6616a2964405590d94d3366)

ccampo133 commented on cyral/resource_cyral_integration_logging.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func (resource *LoggingIntegration) WriteToSchema(d *schema.ResourceData) error {
+
```
```suggestion
```

gabfelp replied:
-- here [35daafa](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/35daafa7a278976ed6616a2964405590d94d3366)

ccampo133 commented on cyral/data_source_cyral_integration_logging.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+func (resp *ListIntegrationLogsResponse) WriteToSchema(d *schema.ResourceData) error {
+	var integrations []interface{}
+	for _, integration := range resp.Integrations {
+		// write in config scheme
+		configType, config, err := getLoggingConfig(&integration)
+		if err != nil {
+			return err
+		}
+
+		integrations = append(integrations, map[string]interface{}{
+			"id":                 integration.Id,
+			"name":               integration.Name,
+			"receive_audit_logs": integration.ReceiveAuditLogs,
+			configType:           config,
+		})
+	}
```
You can avoid unnecessary allocations by allocating the slice ahead of time, since we know its length.

```suggestion
	integrations := make([]interface{}, len(resp.Integrations))
	for i, integration := range resp.Integrations {
		// write in config scheme
		configType, config, err := getLoggingConfig(&integration)
		if err != nil {
			return err
		}

		integrations[i] = map[string]interface{}{
			"id":                 integration.Id,
			"name":               integration.Name,
			"receive_audit_logs": integration.ReceiveAuditLogs,
			configType:           config,
		}
	}
```

gabfelp replied:
Used your proposed solution here: [35daafa](https://github.com/cyralinc/terraform-provider-cyral/pull/395/commits/35daafa7a278976ed6616a2964405590d94d3366). Thank you!

ccampo133 commented on cyral/model_integration_logging.go:
```diff
@@ -0,0 +1,258 @@
+			},
+		},
+
+		"fluentbit": {
```
I would change the name to `fluent_bit` here (and everywhere else it is defined as `fluentbit`).

ccampo133 reviewed (approved):
Added a few comments which should be addressed before merging (plus Wilson's comments), but otherwise approved. Nice work.

ccampo133 commented:
> Retested after the proposed changes and everything is looking fine :)

LGTM @gabfelp
